fpsd/features.py

```diff
@@ -0,0 +1,765 @@
+    """
+    bursts = []
+    current_burst_length = 0
+
+    for row in df.itertuples():
+        if row.ingoing == False:
+            current_burst_length += 1
+        elif row.ingoing == True and current_burst_length > 0:
+            bursts.append(current_burst_length)
+            # Reset burst length
+            current_burst_length = 0
+
+    # If we were in a burst at the end of the trace, let's save it
+    if current_burst_length > 0:
+        bursts.append(current_burst_length)
+
+    # For some features we need to know the ordering of bursts,
+    # so let's save this information as well
+    ranks = list(range(1, len(bursts) + 1))
+    return bursts, ranks
+
+
+class FeatureStorage():
+    def __init__(self, test_db=False):
+        """Set up database engine"""
+        self.TEST_MODE = test_db
+        if self.TEST_MODE:
+            self.engine = create_engine('postgresql://{}:{}@{}/{}'.format(
+                os.environ['PGUSER'], os.environ['PGPASSWORD'],
+                os.environ['PGHOST'], 'test' + os.environ['PGDATABASE']))
+        else:
+            self.engine = create_engine('postgresql://{}:{}@{}/{}'.format(
+                os.environ['PGUSER'], os.environ['PGPASSWORD'],
+                os.environ['PGHOST'], os.environ['PGDATABASE']))
+        return None
+
+    def execute_query_from_string(self, query):
+        """Method to run a query using self.engine
+
+        Args:
+            query [string]: SQL query
+
+        Returns:
+            result [SQLAlchemy ResultProxy object]: result of query
+        """
+
+        result = self.engine.execute(query)
+        return result
+
+    def drop_stale_feature_table(self, table_name):
+        """Try to remove a table even if views depend on it
+
+        Args:
+            table_name [string]: Name of table to be dropped
+        """
+
+        query = "DROP TABLE IF EXISTS {} CASCADE;".format(table_name)
+        self.execute_query_from_string(query)
+        return None
+
+    def generate_table_undefended_frontpage_links(self):
+        """This method creates a table of exampleids that were
+        present at the time of feature generation that the unified
+        view can join on. The table contains one integer column, exampleid.
+        """
+
+        self.drop_stale_feature_table("undefended_frontpage_examples")
+
+        query = ("CREATE TABLE features.undefended_frontpage_examples AS ( "
+                 "SELECT foo.exampleid FROM ( SELECT exampleid, "
+                 "count(*) FROM raw.frontpage_traces GROUP BY exampleid) foo);")
+
+        self.execute_query_from_string(query)
+        return None
+
+    def _create_temp_packet_positions(self, outgoing_only=True):
+        """This method takes all rows in raw.frontpage_traces
+        and creates a temporary table packet_positions
+        with the following format:
+
+        exampleid  | rank     |  ingoing
+        (integer)  | (bigint) |  (boolean)
```

Comment:
> Why create this as a temporary table instead of a view?

The result of a view is not stored, it is basically an alias of a query. When you make several queries on a view, you re-compute the view each time, and then compute the query on that view. It is computationally more efficient to create a temporary table if you are going to be processing multiple queries on the the result of a query.